File: membrain-py/membrain/rerankers/jina.py
"""Jina cross encoder reranker."""

from __future__ import annotations

import time

from ..types import RerankResults, SearchResults
from .base import (
    DEFAULT_TIMEOUT,
    BaseReranker,
    build_rerank_results,
    http_post,
)

_JINA_RERANK_URL = "https://api.jina.ai/v1/rerank"
_JINA_DEFAULT_MODEL = "jina-reranker-v2-base-multilingual"


class JinaReranker(BaseReranker):
    """Cross encoder reranker using the Jina Rerank API.

    Supported models: jina-reranker-v2-base-multilingual, jina-colbert-v2.
    """

    def __init__(
        self,
        api_key: str,
        model: str = _JINA_DEFAULT_MODEL,
        top_k: int = 5,
        endpoint: str | None = None,
        timeout: int = DEFAULT_TIMEOUT,
    ) -> None:
        self.api_key = api_key
        self.model = model
        self.top_k = top_k
        self.endpoint = endpoint or _JINA_RERANK_URL
        self.timeout = timeout

    def rerank(
        self,
        query: str,
        results: SearchResults,
        top_k: int | None = None,
    ) -> RerankResults:
        effective_top_k = top_k if top_k is not None else self.top_k
        memories = results.memories
        if not memories:
            return RerankResults(model=self.model, provider="jina")

        documents = [memory.content for memory in memories]

        start = time.monotonic()
        response = http_post(
            url=self.endpoint,
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json",
            },
            body={
                "query": query,
                "documents": documents,
                "model": self.model,
                "top_n": effective_top_k,
            },
            timeout=self.timeout,
        )
        duration_ms = int((time.monotonic() - start) * 1000)

        scored_indices: list[tuple[int, float]] = []
        for item in response.get("results", []):
            scored_indices.append(
                (item["index"], item["relevance_score"])
            )
        scored_indices.sort(key=lambda pair: pair[1], reverse=True)

        return build_rerank_results(
            memories, scored_indices, self.model, "jina", duration_ms
        )

File: membrain-node/src/rerankers/jina.ts
/**
 * Jina cross encoder reranker.
 */

import type { RerankerConfig, SearchResults, RerankResults } from "../types";

import {
  BaseReranker,
  DEFAULT_TIMEOUT,
  httpPost,
  buildRerankResults,
} from "./base";

const JINA_RERANK_URL = "https://api.jina.ai/v1/rerank";
const JINA_DEFAULT_MODEL = "jina-reranker-v2-base-multilingual";

export class JinaReranker extends BaseReranker {
  private readonly apiKey: string;
  private readonly model: string;
  private readonly topK: number;
  private readonly endpoint: string;
  private readonly timeout: number;

  /**
   * Cross encoder reranker using the Jina Rerank API.
   *
   * Supported models: jina-reranker-v2-base-multilingual, jina-colbert-v2.
   */
  constructor(config: RerankerConfig) {
    super();
    this.apiKey = config.apiKey;
    this.model = config.model ?? JINA_DEFAULT_MODEL;
    this.topK = config.topK ?? 5;
    this.endpoint = config.endpoint ?? JINA_RERANK_URL;
    this.timeout = config.timeout ?? DEFAULT_TIMEOUT;
  }

  async rerank(
    query: string,
    results: SearchResults,
    topK?: number,
  ): Promise<RerankResults> {
    const effectiveTopK = topK ?? this.topK;
    const memories = results.memories;
    if (memories.length === 0) {
      return { memories: [], model: this.model, provider: "jina", duration_ms: 0 };
    }

    const documents = memories.map((memory) => memory.content);

    const start = performance.now();
    const response = await httpPost(
      this.endpoint,
      {
        Authorization: `Bearer ${this.apiKey}`,
        "Content-Type": "application/json",
      },
      {
        query,
        documents,
        model: this.model,
        top_n: effectiveTopK,
      },
      this.timeout,
    );
    const durationMs = Math.round(performance.now() - start);

    const scoredIndices: Array<[number, number]> = [];
    for (const item of response.results ?? []) {
      scoredIndices.push([item.index, item.relevance_score]);
    }
    scoredIndices.sort((a, b) => b[1] - a[1]);

    return buildRerankResults(
      memories,
      scoredIndices,
      this.model,
      "jina",
      durationMs,
    );
  }
}
